// @ionic/core/dist/collection/components/app/app.js
export class App {
    constructor() {
        this.isDevice = false;
        this.deviceHacks = false;
    }
    componentWillLoad() {
        this.isDevice = this.config.getBoolean('isDevice', false);
        this.deviceHacks = this.config.getBoolean('deviceHacks', false);
    }
    hostData() {
        const mode = this.config.get('mode');
        const hoverCSS = this.config.getBoolean('hoverCSS', false);
        return {
            class: {
                [mode]: true,
                'enable-hover': hoverCSS
            }
        };
    }
    render() {
        return [
            this.deviceHacks && h("ion-input-shims", null),
            h("ion-tap-click", null),
            this.isDevice && h("ion-status-tap", null),
            h("slot", null)
        ];
    }
    static get is() { return "ion-app"; }
    static get host() { return { "theme": "app" }; }
    static get properties() { return { "config": { "context": "config" }, "el": { "elementRef": true } }; }
    static get style() { return "/**style-placeholder:ion-app:**/"; }
    static get styleMode() { return "/**style-id-placeholder:ion-app:**/"; }
}
