// @ionic/core/dist/collection/components/segment/segment.js
export class Segment {
    constructor() {
        /*
         * If true, the user cannot interact with the segment. Defaults to `false`.
         */
        this.disabled = false;
    }
    valueChanged(val) {
        this.selectButton(val);
        this.ionChange.emit();
    }
    componentDidLoad() {
        this.selectButton(this.value);
    }
    segmentClick(ev) {
        const selectedButton = ev.target;
        this.value = selectedButton.value;
    }
    selectButton(val) {
        const buttons = this.el.querySelectorAll('ion-segment-button');
        for (let i = 0; i < buttons.length; i++) {
            const button = buttons[i];
            button.activated = (button.value === val);
            // If there is no value set on the segment and a button
            // is checked we should activate it
            if (!val && button.checked) {
                button.activated = button.checked;
            }
        }
    }
    hostData() {
        return {
            class: {
                'segment-disabled': this.disabled
            }
        };
    }
    static get is() { return "ion-segment"; }
    static get host() { return { "theme": "segment" }; }
    static get properties() { return { "color": { "type": String, "attr": "color" }, "disabled": { "type": Boolean, "attr": "disabled" }, "el": { "elementRef": true }, "mode": { "type": "Any", "attr": "mode" }, "value": { "type": String, "attr": "value", "mutable": true, "watchCallbacks": ["valueChanged"] } }; }
    static get events() { return [{ "name": "ionChange", "method": "ionChange", "bubbles": true, "cancelable": true, "composed": true }]; }
    static get style() { return "/**style-placeholder:ion-segment:**/"; }
    static get styleMode() { return "/**style-id-placeholder:ion-segment:**/"; }
}
