// @ionic/core/dist/collection/components/hide-when/hide-when.js
import { updateTestResults, } from '../../utils/show-hide-when-utils';
export class HideWhen {
    constructor() {
        this.orientation = null;
        this.mediaQuery = null;
        this.size = null;
        this.mode = null;
        this.platform = null;
        this.or = false;
        this.passesTest = false;
    }
    componentWillLoad() {
        return updateTestResults(this);
    }
    hostData() {
        return {
            class: {
                'show-content': !this.passesTest,
                'hide-content': this.passesTest
            }
        };
    }
    static get is() { return "ion-hide-when"; }
    static get properties() { return { "calculatedPlatforms": { "context": "platforms" }, "config": { "context": "config" }, "element": { "elementRef": true }, "mediaQuery": { "type": String, "attr": "media-query" }, "mode": { "type": String, "attr": "mode" }, "or": { "type": Boolean, "attr": "or" }, "orientation": { "type": String, "attr": "orientation" }, "passesTest": { "state": true }, "platform": { "type": String, "attr": "platform" }, "size": { "type": String, "attr": "size" } }; }
    static get style() { return "/**style-placeholder:ion-hide-when:**/"; }
}
