// @ionic/core/dist/collection/components/range-knob/range-knob.js
export class RangeKnob {
    handleKeyBoard(ev) {
        const keyCode = ev.keyCode;
        if (keyCode === KEY_LEFT || keyCode === KEY_DOWN) {
            this.ionDecrease.emit({ isIncrease: false, knob: this.knob });
            ev.preventDefault();
            ev.stopPropagation();
        }
        else if (keyCode === KEY_RIGHT || keyCode === KEY_UP) {
            this.ionIncrease.emit({ isIncrease: true, knob: this.knob });
            ev.preventDefault();
            ev.stopPropagation();
        }
    }
    leftPos(val) {
        return `${val * 100}%`;
    }
    hostData() {
        return {
            class: {
                'range-knob-pressed': this.pressed,
                'range-knob-min': this.val === this.min || this.val === undefined,
                'range-knob-max': this.val === this.max
            },
            style: {
                'left': this.leftPos(this.ratio)
            },
            'role': 'slider',
            'tabindex': this.disabled ? -1 : 0,
            'aria-valuemin': this.min,
            'aria-valuemax': this.max,
            'aria-disabled': this.disabled,
            'aria-labelledby': this.labelId,
            'aria-valuenow': this.val
        };
    }
    render() {
        if (this.pin) {
            return [
                h("div", { class: 'range-pin', role: 'presentation' }, this.val),
                h("div", { class: 'range-knob', role: 'presentation' })
            ];
        }
        return h("div", { class: 'range-knob', role: 'presentation' });
    }
    static get is() { return "ion-range-knob"; }
    static get properties() { return { "disabled": { "type": Boolean, "attr": "disabled" }, "knob": { "type": String, "attr": "knob" }, "labelId": { "type": String, "attr": "label-id" }, "max": { "type": Number, "attr": "max" }, "min": { "type": Number, "attr": "min" }, "pin": { "type": Boolean, "attr": "pin" }, "pressed": { "type": Boolean, "attr": "pressed" }, "ratio": { "type": Number, "attr": "ratio" }, "val": { "type": Number, "attr": "val" } }; }
    static get events() { return [{ "name": "ionIncrease", "method": "ionIncrease", "bubbles": true, "cancelable": true, "composed": true }, { "name": "ionDecrease", "method": "ionDecrease", "bubbles": true, "cancelable": true, "composed": true }]; }
}
export const KEY_LEFT = 37;
export const KEY_UP = 38;
export const KEY_RIGHT = 39;
export const KEY_DOWN = 40;
